Use nucleo for fuzzy string matching

P1n3appl3 opened the pull request:
Closes #7586

Release Notes:

- Fuzzy pickers now use the [nucleo](https://docs.rs/nucleo/latest/nucleo) library from the [Helix editor](https://helix-editor.com) TODO: add a note about new syntax and/or some examples

P1n3appl3 committed:
Use nucleo for fuzzy string matching

Co-authored-by: David Kleingeld <davidsk@zed.dev>

dionysuzx commented:
will this enable fuzzy search like vscode/telescope? currently if there's a file in "someDir/file.md", and i search "file somedir" it won't find it. it only worked if i respect the order like "somedir file".

i feel addressing the above usage is extremely important, as it's related to file search; and i'm really surprised it hasn't been an issue for others (unless i'm doing something wrong). thanks! :)

dvdsk commented:
> will this enable fuzzy search like vscode/telescope? currently if there's a file in "someDir/file.md", and i search "file somedir" it won't find it. it only worked if i respect the order like "somedir file".

It will :partying_face: (when done, we still need to do the path bit)

P1n3appl3 committed:
Add fuzzy path matching

Add path match test

Fix segment splitting

Co-authored-by: David Kleingeld <davidsk@zed.dev>

Add tests and tweak ordering

Co-authored-by: David Kleingeld <davidsk@zed.dev>

Fix penalize_length

Fix all the Ord impls

Co-authored-by: David Kleingeld <davidsk@zed.dev>

Fix file_finder ordering

Co-authored-by: David Kleingeld <davidsk@zed.dev>

Add tests

dvdsk commented:
status update:
We have the implementation done and are now adding tests.
    
Todo
 - Check the case that kirill's comment talked about in file_finder::matching_history_items
 - Write a test case for smit's completion fuzzy matching (had something to do with prioritizing prefixes or long strings)
 - Run on the zed source tree to make sure it works
 - Hunt down the bug where some paths aren't showing up whatsoever in fuzzy path search results

P1n3appl3 committed:
wip

P1n3appl3 committed:
Weight path matches differently

Co-authored-by: David Kleingeld <davidsk@zed.dev>

P1n3appl3 committed:
Fix smart_case

P1n3appl3 committed:
Fix a bunch of things

Co-authored-by: David Kleingeld <davidsk@zed.dev>

P1n3appl3 committed:
Format

P1n3appl3 commented:
We've fixed a bunch more tests, and now we're at the point where the remaining failures seem to be mostly discrepancies between our expectations and nucleo's scoring algorithm. We've opened helix-editor/nucleo#92 to check with upstream about one of those which impacts code completion tests pretty heavily.
